### ISSUE TITLE
Simplify SSL write API for Mbed TLS 3

### ISSUE THREAD

**Issue by hanno-arm** [ARMmbed/mbedtls#4183]
__Context:__ The SSL API `mbedtls_ssl_write()` is for writing application data through an established TLS connection. In Mbed TLS 2.X, the semantics of this API is that it always attempts to deliver the application data to the underlying transport immediately, and returning a dedicated non-fatal error `MBEDTLS_ERR_SSL_WANT_WRITE` if that's not possible. 

__Problem:__ There are two issues arising from this design:
1. It is not possible to batch multiple calls to `mbedtls_ssl_write()` prior to delivering them to the underlying transport. This is arguably not an issue for TLS because the user could concatenate the data itself before calling `mbedtls_ssl_write()` _once_, but it is a limitation for DTLS, where every invocation of `mbedtls_ssl_write()` shall correspond to exactly one DTLS datagram, multiple of which may be delivered in the same datagram of the underlying transport (such as UDP).
2. More importantly, when calling `mbedtls_ssl_write()`, we have to distinguish the following situations:
    * Data prepared in an earlier call to `mbedtls_ssl_write()` is still to be delivered but the underlying transport is not available, returning `MBEDTLS_ERR_SSL_WANT_WRITE`. The new data was not prepared queued for sending. 
    * The new data was prepared and queued for sending, but couldn't be delivered because the underlying transport is not available, again returning `MBEDTLS_ERR_SSL_WANT_WRITE`. 
   
     With the current design, it is impossible to tell those cases apart. The resolution employed by Mbed TLS 2.X is to mandate that if `mbedtls_ssl_write()` fails with `MBEDTLS_ERR_SSL_WANT_WRITE`, then the user _must_ retry the call with the exact same parameters until `mbedtls_ssl_write()` succeeds, and only then it may be called again with different data. In fact, the implementation entirely ignores the arguments to `mbedtls_ssl_write()` if an earlier write is still unfinished, and `mbedtls_ssl_write()` is demoted to a plain flush in this case -- see [here](https://github.com/ARMmbed/mbedtls/blob/34045c1d6a28cd7c8d6e4b43c4dcbf0f4f66b513/library/ssl_msg.c#L5783-5814)

In addition to not allowing batching of datagrams, the above design seems confusing, and we've had issues with it in the past.

__Task:__ Agree on and implement a more suitable writing API.

__Concrete proposal:__ 
Split the writing API in two parts: 
1. `mbedtls_ssl_write()` which prepares outgoing application data for sending and queues it internally, but does not yet send it.
2. `mbedtls_ssl_flush()`, which attempts to deliver all data previously prepared via `mbedtls_ssl_write()` to the underlying transport.
Both calls may fail with `MBEDTLS_ERR_SSL_WANT_WRITE`, but for different reasons:
* If `mbedtls_ssl_write()` fails with `MBEDTLS_ERR_SSL_WANT_WRITE`, it's because new outgoing data can only be prepared after previously prepared data has been completely or partially flushed, freeing  up internal space. In this case, the user knows that its data was _not_ prepared and queued, but there is no obligation to call `mbedtls_ssl_write()` again with the same or different parameters - it's up to the user.
* If `mbedtls_ssl_flush()` fails with `MBEDTLS_ERR_SSL_WANT_WRITE`, then not all data previously prepared via `mbedtls_ssl_write()` could be delivered to the underlying transport.

This approach fixes both issues above: At no point is the user obliged to call `mbedtls_ssl_write()` again with specific parameters, and it's always unambiguous which data has successfully been passed to the TLS stack and only requires flushing, and for which one the user should call `mbedtls_ssl_write()` again. Moreover, the design allows for multiple DTLS datagrams to be packed into a single UDP datagram by the messaging implementation, because `mbedtls_ssl_write()` is no longer forced to attempt immediate delivery.

**Comment by hanno-arm** [comment]
A similar change should likely be made to `mbedtls_ssl_close_notify()`.

**Comment by gilles-peskine-arm** [comment]
> Split the writing API in two parts:

What is the migration path for existing code? Is the old `mbedtls_ssl_write(ssl, buf, size)` equivalent to this?
```
write_ret = mbedtls_ssl_write(ssl, buf, size);
if (write_ret < 0) return write_ret;
flush_ret = mbedtls_ssl_flush(ssl);
if (flush_ret < 0) return flush_ret;
return write_ret;
```

> If mbedtls_ssl_flush() fails with MBEDTLS_ERR_SSL_WANT_WRITE, then not all data previously prepared via mbedtls_ssl_write() could be delivered to the underlying transport.

I don't understand in what case `mbedtls_ssl_flush` would return `WANT_WRITE` rather than some other error, and what the corrective action is. Attempting to write to a non-blocking socket when it would block, maybe?

**Comment by hanno-arm** [comment]
@gilles-peskine-arm Yes, you got the migration path right. I'm currently drafting an implementation of the new API, and for the example programs, it comes down to adding an `mbedtls_ssl_flush()` loop at the end of the `mbedtls_ssl_write()` loop. The `mbedtls_ssl_write()` loop itself need not be changed, though.

> I don't understand in what case mbedtls_ssl_flush would return WANT_WRITE rather than some other error

I'm not sure I got your question right, but `mbedtls_ssl_flush()` returns `WANT_WRITE` if not all data previously dispatched via `mbedtls_ssl_write()` could be delivered to the underlying transport.

**Comment by hanno-arm** [comment]
> What is the migration path for existing code? Is the old mbedtls_ssl_write(ssl, buf, size) equivalent to this? 

@gilles-peskine-arm I was a bit inaccurate above: While the pattern is right, the code you gave will _not_ be equivalent to the old semantics of `mbedtls_ssl_write()`. Instead, it would be something like
```c
if( write_in_progress == 0 )
{
    write_in_progress = 1;
    write_done = 0;
}

if( write_done == 0 )
{
    write_ret = mbedtls_ssl_write( ssl, buf, size );
    if( write_ret < 0 )
        return( write_ret );
    write_done = 1;
}

flush_ret = mbedtls_ssl_flush( ssl );
if( flush_ret != 0 )
    return( flush_ret );

write_in_progress = 0;
return( write_ret );
```

**Comment by hanno-arm** [comment]
@mpg @gilles-peskine-arm Any chance to get this in, or to we have to call it out of scope?

**Comment by jethrogb** [comment]
Until the new API is in place, are you willing to commit to the following API description change? I believe it doesn't require any code changes.

`mbedtls_ssl_flush_output` is now considered public API. If `mbedtls_ssl_write` returns `MBEDTLS_ERR_SSL_WANT_WRITE`, then instead of calling `mbedtls_ssl_write` again, you may call `mbedtls_ssl_flush_output`. If you do so, you must call `mbedtls_ssl_flush_output` until it returns 0. Now you may call `mbedtls_ssl_write` again with different parameters.

**Comment by dragan-ponos** [comment]
I think this issue is very important for the logic I am writing. Could you please assist me for the following scenario?
1. I try to **write** data, but it returns `MBEDTLS_ERR_SSL_WANT_WRITE`.
2. The upper layer decides to **close** the connection.

I would like to **"abort" the write** and proceed with sending **close_notify**.
Is this possible in the current API? Will it be possible in the upcoming API?
I get the impression it is not, and I would need to write data fully before proceeding with close_notify.
Thanks!

**Comment by gilles-peskine-arm** [comment]
@dragan-ponos Thanks for the feedback! Indeed on code inspection this looks like a bug in `mbedtls_ssl_close_notify`: I guess it should return `WANT_WRITE` if `flush_output` returns 0? Would you mind filing a separate issue with at least rough instructions to reproduce the problem, since this seems like a bug we could fix independently of an API change?
